### ISSUE TITLE
Defining CVs and Biases with an external software and apply it to the OpenMM simulation

### ISSUE THREAD

**Issue by DanielWicz** [openmm/openmm#2911]
In the big picture, I would like to have access to defining CVs and bias through external Python software and be able to apply it to the simulation by means of two callable objects. Let's define those two callable Python objects `newCV` and `newBias` from an external Python package. Those two callable objects accept coordinates of a molecule (e.g. drug, protein), further, `newBias` accepts CV, produces by the first object (`newCV`). Thus a conceptual code would look like:

```
# Here is my package
from mypackage import newCV, newBias

# Copy from OpenMM documentation starts here
from simtk.openmm.app import *
from simtk.openmm import *
from simtk.unit import *
from sys import stdout, exit, stderr

psf = CharmmPsfFile('input.psf')
pdb = PDBFile('input.pdb')
params = CharmmParameterSet('charmm22.rtf', 'charmm22.prm')
system = psf.createSystem(params, nonbondedMethod=NoCutoff,
        nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinIntegrator(300*kelvin, 1/picosecond, 0.002*picoseconds)
simulation = Simulation(psf.topology, system, integrator)
simulation.context.setPositions(pdb.positions)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter('output.pdb', 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True,
        potentialEnergy=True, temperature=True))

# Copy from OpenMM documentation ends here
# 10000 is number of steps
for step in range(10000):
    # Here I use some conceptual methods, they don't exist in reality (except step) 
    # Next lines, that start with ##, describe those conceptual methods
    ## Here I get coordinates from a selection
    coord = simulation.getCoords(selection='protein and type Calpha')
    ## I get new CVs (some distance between atoms **_n_** and **_m_** or angles between atoms
    ## **_n_**, **_m_**,  **_g_**
    newpredictedCV = newCV(coordinates=coord)
    ## Get biases for the new CVs, where the bias is predicted for a given
    ## point in space (e.g. **_a_**, **_b_**, **_c_**) for three CVs in the CV's coordinate space
    newpredictedBias = newBias(CV=newpredictedCV, coord)
    
    ## apply the CVs and then Biases to the simulation
    simulation.applyCV(newpredictedCV).applyBias(newpredictedBias)
    ## perform one (or more steps) with the new bias and CVs
    ## most probably more than one, to let the CVs diffuse
    simulation.step(1)
```
1. It is possible to extract coordinates through a selection in openMM?
2. It is possible to apply a CV in openMM
3. it is possible to apply external bias in openMM ?

**Comment by peastman** [comment]
That isn't really how OpenMM works.  You don't call a function to "apply" a force to the system.  You add Force objects to the system that define how the calculation is implemented.  You could hack something together to do this (call `getState()` to retrieve the coordinates, pass them to your code, then store the forces into per-particle parameters on a CustomExternalForce that applies forces based on them), but that will be very inefficient.

What about using the [TensorFlow](https://github.com/openmm/openmm-tensorflow) or [PyTorch](https://github.com/openmm/openmm-torch) plugin to implement your CV and bias calculation?

Another option, of course, is to write a plugin that implements your new Force class.

**Comment by DanielWicz** [comment]
> 
> 
> That isn't really how OpenMM works. You don't call a function to "apply" a force to the system. You add Force objects to the system that define how the calculation is implemented. 
Sorry for the inconvenience, it is my first time using OpenMM, usually, I was using the other two big MD frameworks.

>You could hack something together to do this (call `getState()` to retrieve the coordinates, pass them to your >code, then store the forces into per-particle parameters on a CustomExternalForce that applies forces based on >them), but that will be very inefficient.
Does it update all system's forces or only those from some group?

> What about using the [TensorFlow](https://github.com/openmm/openmm-tensorflow) or [PyTorch](https://github.com/openmm/openmm-torch) plugin to implement your CV and bias calculation?
> 
> Another option, of course, is to write a plugin that implements your new Force class.
More/less What I'm intended to do and actually and  I saw those plugins already. Nevertheless, in my case I would like to interact with the system directly without saving a TensorFlow/PyTorch model. So that I update the TF/PyTorch model depending on the current state of the MD system. So the schema would be:
1. Perform MD simulation
2. Put some atom group into my algorithm (e.g. protein or maybe ligand)
3. The algorithm outputs current CVs and Bias for the CVs as described before
4. I calculate some metric
5. I update the Tensorflow algorithm
6. Repeat from 1 to 5, until the user terminates or some other metric converges

>Another option, of course, is to write a plugin that implements your new Force class.
Wouldn't it be possible also by using `CustomCVForce` or by modyfing e.g. `Metadynamics` class or `BiasVariable`?
Consider it that, I'm not an expert in C++ and I wan't to avoid it, but if it will be necessary then okay.

**Comment by peastman** [comment]
Can you give more details on exactly what you want to calculate?  In particular,

- What are the CVs you want to calculate?
- How does the bias force depend on the CVs?
- How does each of those change during the simulation?

There are various ways you could try to implement this, all involving different tradeoffs.  The details of your calculation will determine which is best.

**Comment by DanielWicz** [comment]
>Can you give more details on exactly what you want to calculate? In particular,

>-     What are the CVs you want to calculate?
CVs are some kind of periodic function, e.g. backbone's angles in dialanine. Yet, I don't assume which angles are chosen, but predict them with `newCV(coordinates=coord)` based on a subset of coordinates (e.g. dialanine backbone).  The number of predicted CVs is N, where N is chosen by the user (e.g. 2) and CVs are predicted by some sort of neural network during the simulation. So for example from the list of ~10 possible CVs, three are going to be predicted every cycle (cycle is explained better at the end) of the simulation ( for example as an one-hot encoding matrix).

>- How does the bias force depend on the CVs?
Since I want to keep it now "as simple as possible", thus I'm going to use Fourier basis for the periodic CVs. Although the basis's coefficients are going to be predicted by some sort of neural network, which has coordinates (similarly to `newCV(coordinates=coord)`) and predicted CV as an input (i.e. `newBias(CV=newpredictedCV, coord)`).

>- How does each of those changes during the simulation?
Let me show it as a step-form:
0. Minimze energy of the system (as usually in MD simulations).
    1. Predict N CVs based on the current coordinates CR
    2. Predict basis's coefficients based on the current coordinates CR for each of N CVs.
    3. Perform biased MD simulation (for example 5000 integration steps) with the predicted CVs and bias.
    4. Output estimated free energy estimate along predicted CVs and MD simulation.
    5. Calculate some metric and update ANN responsible for CVs and bias's basis coefficients.
6. Repeat 1-4 for M cycles, until used termination.

So generally metadynamics without KDE, gaussian deposition and non-constant CVs, if I'm not mistaken.

**Comment by peastman** [comment]
There are various ways you could implement this.  All of them will have inefficiencies of various sorts, but I think the following will be simplest overall.

1. Use a CustomTorsionForce to implement the bias.  If I understand correctly, the bias potential is just a simple function of the angle, so a single force object is all you need.  More complicated arrangements could be implemented with a CustomCompoundBondForce and/or CustomCVForce, but I don't think you need that.

2. Every 5000 steps (or whatever it is), update the CustomTorsionForce.  It defines N torsions, and you need to call `setTorsionParameters()` to update each of them.  If all you're changing is the per-torsion parameters, you can then call `updateParametersInContext()` on it, which is pretty fast.  But if you're also changing which atoms they're defined by, that requires calling `reinitialize()` on the Context, which is much more expensive.  Given everything else you're doing, it hopefully still won't slow things down too much.

**Comment by DanielWicz** [comment]
> 
> 
> There are various ways you could implement this. All of them will have inefficiencies of various sorts, but I think the following will be simplest overall.
> 
>     1. Use a CustomTorsionForce to implement the bias.  If I understand correctly, the bias potential is just a simple function of the angle, so a single force object is all you need.  More complicated arrangements could be implemented with a CustomCompoundBondForce and/or CustomCVForce, but I don't think you need that.
> 
>     2. Every 5000 steps (or whatever it is), update the CustomTorsionForce.  It defines N torsions, and you need to call `setTorsionParameters()` to update each of them.  If all you're changing is the per-torsion parameters, you can then call `updateParametersInContext()` on it, which is pretty fast.  But if you're also changing which atoms they're defined by, that requires calling `reinitialize()` on the Context, which is much more expensive.  Given everything else you're doing, it hopefully still won't slow things down too much.

Thanks a lot, Peter I will try to test it. With regard to changing parameter (i.e. weights of fourier's expansion), I set them as global parameters using `addGlobalParameter()` and change with `setParameter()` for every cycle ?

**Comment by peastman** [comment]
That's fine.  If the number of parameters is small enough that you can use global parameters, those are very fast to update.

**Comment by peastman** [comment]
Closing this since I think everything has been answered.  If any part of this hasn't been resolved yet, feel free to reopen it.
